File: aws-ec2/operation/create_security_group.rb
=begin
##########################################################################
#
#  INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
#  __________________
# 
#  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
#  All Rights Reserved.
#  Product / Project: Flint IT Automation Platform
#  NOTICE:  All information contained herein is, and remains
#  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
#  The intellectual and technical concepts contained
#  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
#  Dissemination of this information or any form of reproduction of this material
#  is strictly forbidden unless prior written permission is obtained
#  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
=end

# begin
@log.trace("Started executing 'fb-cloud:aws-ec2:operation:create_security_group.rb' flintbit...")
# Flintbit Input Parameters
# Mandatory
connector_name = @config.global('flintcloud-integrations.aws-ec2.name')	      # Name of the Amazon EC2 Connector
action = 'create-security-group'                    # Specifies the name of the operation: create-security-group
group_name = @input.get('group_name')	              # Contain security group name that you want to create
group_description = @input.get('group_description')	# Contain security group description that you want to create
# Optional

region = @input.get('region')	                      # Amazon EC2 region (default region is 'us-east-1')
request_timeout = @input.get('timeout')	            # Execution time of the Flintbit in milliseconds (default timeout is 60000 milloseconds)

@log.info("Flintbit input parameters are, action : #{action} | group_name : #{group_name} | region : #{region} | access_key : #{@access_key} | secret_key : #{@secret_key} ")

connector_call = @call.connector(connector_name).set('action', action).set('group-name', group_name).set('group-description', group_description)
                      .set("access-key",@input.get("access-key"))
                          .set("security-key",@input.get("security-key"))
if !region.nil? && !region.empty?
    connector_call.set('region', region)
else
    @log.trace("region is not provided so using default region 'us-east-1'")
end

if request_timeout.nil? || request_timeout.is_a?(String)
    @log.trace("Calling #{connector_name} with default timeout...")
    response = connector_call.sync
else
    @log.trace("Calling #{connector_name} with given timeout #{request_timeout}...")
    response = connector_call.timeout(request_timeout).sync
end

# Amazon EC2 Connector Response Meta Parameters
response_exitcode = response.exitcode	# Exit status code
response_message = response.message	# Execution status messages

if response_exitcode == 0
    @log.info("SUCCESS in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
    @user_message = "Successfully created AWS Security Group"
     @output.set('message', response.message).set('exit-code',0).set('user_message',@user_message)
else
    @log.error("ERROR in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
    response=response.to_s
    @user_message = "Failed to create AWS Security Group"
    if !response.empty?
    @output.set('message', response_message).set('exit-code', 1).setraw('error-details',response.to_s).set('user_message',@user_message)
    else
    @output.set('message', response_message).set('exit-code', 1).set('user_message',@user_message)
    end
end
@log.trace("Finished executing 'fb-cloud:aws-ec2:operation:create_security_group.rb' flintbit")
# end

File: aws-ec2/operation/workflow/create_security_group.js
/**
 * Creation Date: 28/05/2019
 * Summary: To create security group on AWS
 * Description: To create security group using aws-ec2 connector
 */

log.trace("Started executing the flintbit 'fb-cloud:aws-ec2:operation:workflow:create_security_group.js'");

input_scope = JSON.parse(input);   //For checking if the input contains the necessary keys

//Connector Name  - mandatory
connector_name = "amazon-ec2";
connector_call = call.connector(connector_name);
log.info("Connector Name: "+connector_name);

//Action - mandatory
action = "create-security-group";
log.info("Action: "+action);

connector_call.set("action",action);

if(input_scope.hasOwnProperty("cloud_connection")){

    //Access Key - mandatory
    access_key = input_scope.cloud_connection.encryptedCredentials["access_key"];
    if(access_key!=null || access_key!=""){
       connector_call.set("access-key",access_key); 
       log.info("Access Key is given");
    }
    else{
        log.error("Access-Key is null or empty string.");
    }

    //Security Key - mandatory
    secret_key = input_scope.cloud_connection.encryptedCredentials["secret_key"];
    if(secret_key!=null || secret_key!=""){
       connector_call.set("security-key",secret_key); 
       log.info("Security Key is given");
    }
    else{
        log.error("Security-Key is null or empty string.");
    }

    //Group Name - mandatory
    if(input_scope.hasOwnProperty("group_name")){
        group_name = input.get("group_name");
        if(group_name!=null || group_name!=""){
            connector_call.set("group-name",group_name); 
            log.info("Group Name: "+group_name);
        }
        else{
            log.error("Group Name is null or empty string.")
        }
    }
    else{  //group_name key not present in input JSON 
        log.error("Input does not contain the key 'group_name'")
    }

    //Group Discription - mandatory
    if(input_scope.hasOwnProperty("group_description")){
        group_description = input.get("group_description");
        if(group_description!=null || group_description!=""){
            connector_call.set("group-description",group_description); 
            log.info("Group Description: "+group_description);
        }
        else{
            log.error("Group Description is null or empty string.")
        }
    }
    else{  //group_description key not present in input JSON 
        log.error("Input does not contain the key 'group_description'")
    }

    //Region - mandatory
    if(input_scope.hasOwnProperty("region")){
        region = input.get("region");
        if(region!=null || region!=""){
            connector_call.set("region",region); 
            log.info("Region: "+region);
        }
        else{
            log.error("Region is null or empty string.");
        }
    }
    else{  //region key not present in input JSON 
        log.error("Input does not contain the key 'region'");
    }

    //Timeout - NOT mandatory
    if(input_scope.hasOwnProperty("request_timeout")){
        request_timeout = input.get("request_timeout");
        if(request_timeout!=null || request_timeout!=""){
            connector_call.set("timeout",request_timeout); 
            log.info("Request Timeout: "+request_timeout);
        }
        else{
            connector_call.set("timeout",240000); 
            log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
        }
    }
    else{
        connector_call.set("timeout",240000); 
        log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
    }

    //Connector call
    response = connector_call.sync();

    //Response meta-parameters
    response_exitcode = response.exitcode();
    response_message = response.message();

    //Response Result
    group_id = response.get("group-id");

    if(response_exitcode==0){
        log.info("Successfully created a Security Group");
        user_message = "Successfully created a Security Group with <b>Group-ID: "+group_id+"</b>";
        output.set("message",response_message)
            .set("exit-code",response_exitcode)
            .set("user_message",user_message)
            .set("result",group_id);
        log.trace("Finished executing 'fb-cloud:aws-ec2:operation:workflow:create_security_group.js' successfully");
    }
    else{
        log.error("Failure in execution, message:"+response_message+" | exitcode:"+response_exitcode);
        output.set("error",response_message).set("exit-code",-1);
        log.trace("finished executing 'fb-cloud:aws-ec2:operation:workflow:create_security_group.js' with errors")
    }
}
else{ //cloud_connection key not present in input JSON
    log.error("Cloud Connection not given. Can not authenticate without Secret-Key and Access-Key");
}